### PR TITLE
 README: add usage for nixos-18.09 

### PR DESCRIPTION
**Pull request by colemickens**
@zimbatm can you test out these instructions, assuming you have an 18.09 system?

rendered view: https://github.com/colemickens/nixpkgs-wayland/blob/usage-18.09/README.md see the "usage" section

### CODE REVIEW

**Comment by zimbatm** [comment]
`programs.sway.package` isn't available in 18.09, investigating

**Comment by colemickens** [comment]
Yeah, I added the `package` option to the module and didn't backport it (I've not really backported anything yet and don't know what the bar for backports is, etc.)

I'm guessing there's probably a more appropriate way to do this so that it would override the original package, so that setting the option wouldn't be necessary..,

**Comment by zimbatm** [comment]
This should do the trick, testing:
```
{ ... }:
{
  nixpkgs.overlays = [
    (self: super: {
     sway = super.swaypkgs.sway-beta;
   })
];
```

**Comment by zimbatm** [comment]
NOTE: `programs.sway.extraPackages` might be a better place to put all the sway dependencies

**Comment by colemickens** [comment]
>(I am also testing swaypkgs as an overlay for now to see if anything breaks)

Hm, what does that mean? Since `swaypkgs` is a full nixpkgs tree, wouldn't that sort of pull all of your packages up to `nixos-unstable` effectively?

>NOTE: programs.sway.extraPackages might be a better place to put all the sway dependencies

Agreed, I'll do that before merging.

File: README.md [hunk 1]
```diff
@@ -33,26 +33,81 @@ Please open an issue if something is out of date.
...
-    then (import /etc/nixpkgs-wayland)
-    else (import (builtins.fetchTarball nos));
+  swayOverlay = (import (builtins.fetchTarball nos));
+  nixpkgs = import (builtins.fetchTarball {
+    url = "https://github.com/NixOS/nixpkgs-channels/archive/nixos-unstable.tar.gz";
+  });
+  swaypkgs = (import nixpkgs {
+    config = config.nixpkgs.config;
+    overlays = [ swayOverlay ];
+  }).pkgs;
```
Comment (zimbatm) [inline README.md]: A simpler version, this has the advantage that swaypkgs controls the nixpkgs version and so should be more compatible with the binary-cache:
```suggestion
  swaypkgs = import "${builtins.fetchTarball nos}/build.nix";
```

File: README.md [hunk 2]
```diff
@@ -33,26 +33,81 @@ Please open an issue if something is out of date.
...
-
-  environment.systemPackages = with pkgs; [
-    sway-beta grim slurp wlstream
-    redshift-wayland waybar
-  ];
-}
+  {
+    programs.sway = {
+      enable = true;
+      package = swaypkgs.sway-beta;
```
Comment (zimbatm) [inline README.md]:
```suggestion
```

File: README.md [hunk 3]
```diff
@@ -33,26 +33,81 @@ Please open an issue if something is out of date.
...
+    };
```
Comment (zimbatm) [inline README.md]:
```suggestion
    };
    nixpkgs.overlays = [ (self: super: { sway = swaypkgs.sway-beta; }) ];
```

**Comment by zimbatm** [comment]
> Hm, what does that mean? Since swaypkgs is a full nixpkgs tree, wouldn't that sort of pull all of your packages up to nixos-unstable effectively?

nevermind that, I used the proposed changes instead in the end.

The system builds, I just need to reboot and test the runtime properly tomorrow

**Comment by colemickens** [comment]
Okay, I think I see what you've done and I expected it will work. One question -- I've been writing up instructions on how this can actually be used on Ubuntu with Nix, and in pursuit of that, I've been building against `nixos-unstable` and `nixpkgs-unstable`. I'm just not sure how to make this work cleanly other than maybe relying on `<nixpkgs>` which is unpredictable, impure IMO, and is even less ideal in terms of binary cache hits.

Let me know if you have any thoughts or if you think I should just not bother with building against both channels, or if there's another nice solution...

**Comment by zimbatm** [comment]
in terms of channels, `nixos-unstable` is the same as `nixpkgs-unstable` but advances from master on a different set of criteria. It's generally better tested as it's also running integration tests for some of the services.

Based on that I would only publish one channel with a nixkgs version pinned as part of the distribution. The main difference for Ubuntu would be the installation instructions.

**Comment by colemickens** [comment]
Okay, I see. Any word on if it worked for you when you tried it?

**Comment by colemickens** [comment]
* Yeah, the `sway` devs seem opinionated about their lack of support for login managers. I haven't really bothered testing with it, and for my "Sway on Ubuntu" instructions, I have the user "isolate" back to the TTY before starting sway.
* Hm, I don't have this problem with NetworkManager, but maybe it's because it's storing it in /etc instead of my keychain?
* RE: `sway` config - Is this something I can/should fix? I don't totally understand how default config files are supposed to be handled for applications OR modules in NixOS.
* RE: `waybar` config - again, is this something I can fix in the waybar package or is it just a general documentation thing?
* I packaged something called `kanshi` that might be useful... I haven't tested it out yet but it's there and ready to go and is meant to help with dynamic multi-monitor situations.

**Comment by zimbatm** [comment]
1. The package includes a `sway.desktop` file in `share/wayland-sessions` so in theory it could work. The [Arch wiki also mentions that it's unsupported](https://wiki.archlinux.org/index.php/Sway#From_a_display_manager).
2. Yeah this is because I am coming from KDE who includes it's own "kwallet" keychain. NetworkManager is probably falling back on a file store solution.
3. It could be mentioned in the doc. A `sway-config-init` script could be provided for convenience's sake
4. Just a doc thing
5. Awesome, I will check it out tomorrow

Thanks for all the hard work! I think this PR can be merged as those are just future potential improvements. If you want me to split some into tickets let me know.

**Comment by colemickens** [comment]
I think I need to implement the change you suggested on GitHub. The way GH is showing it to me makes it hard to parse, I'll take a closer look tomorrow and merge it. Thanks for testing and the insight!

**Comment by bbigras** [comment]
I'm trying to use the instruction with 18.09 but I got:

```
error: syntax error, unexpected WITH, at /etc/nixos/configuration.nix:100:7
(use '--show-trace' to show detailed location information)
building Nix..
```

It's about the `with swaypkgs` 'with' after the `++`. I did a copy paste. I'm in #sway on freenode. My nick should be 'bbigras'.

File: README.md [hunk 4]
```diff
@@ -33,26 +33,81 @@ Please open an issue if something is out of date.
...
+    environment.systemPackages =
+      with pkgs; [
+        vim git
+      ] ++
+      with swaypkgs; [
```
Comment (zimbatm) [inline README.md]:
```suggestion
      (with swaypkgs; [
```

File: README.md [hunk 5]
```diff
@@ -33,26 +33,81 @@ Please open an issue if something is out of date.
...
+        # DO NOT list `sway-beta` here! It will override the module!
+        # sway-beta
+
+        # essentials for any wlroots compositor
+        grim slurp mako wlstream redshift-wayland
+
+        # optional alternative bar
+        waybar
+      ];
```
Comment (zimbatm) [inline README.md]:
```suggestion
      ]);
```

**Comment by zimbatm** [comment]
@bbigras ^^

**Comment by bbigras** [comment]
Thanks! I was able to `nixos-rebuild` but I got this error when I try to start sway:
`Couldn't exec sway-setcap: Permission denied`

configuration.nix
```nix
let
  nos = "https://github.com/colemickens/nixpkgs-wayland/archive/master.tar.gz";
  swayOverlay = (import (builtins.fetchTarball nos));
  nixpkgs = import (builtins.fetchTarball {
    url = "https://github.com/NixOS/nixpkgs-channels/archive/nixos-unstable.tar.gz";
  });
  swaypkgs = import "${builtins.fetchTarball nos}/build.nix";
in
{
# [...]
  programs.sway = {
    enable = true;
  };

  nixpkgs.overlays = [ (self: super: { sway = swaypkgs.sway-beta; }) ];

    environment.systemPackages =
      with pkgs; [
      ] ++
      (with swaypkgs; [
        # DO NOT list `sway-beta` here! It will override the module!
        # sway-beta

        # essentials for any wlroots compositor
        grim slurp mako wlstream redshift-wayland

        # optional alternative bar
        waybar
      ]);
```

**Comment by zimbatm** [comment]
it might also be required to add the logged-in user to the "sway" group. This will require a logout/login to be active.

```nix
{
  users.extraUsers.youruser = {
    extraGroups = [ "sway" ];
  };
}
```

**Comment by colemickens** [comment]
You both want to make sure you're using 'sway-beta' for the package name
AND the module name. I keep seeing 'programs.sway' (which may be what I had
originally, but it's wrong).

**Comment by zimbatm** [comment]
in that case the module would have to be imported as well since it's not available in 18.09

**Comment by colemickens** [comment]
Oh okay, I see now that `sway` is being set to `sway-beta` from the overlay, so that should actually work.

I might just go ahead and drop `-beta` from the package name as well, we'll see.

**Comment by colemickens** [comment]
@zimbatm I've rewritten more of this to take into account your suggestions and further simplify it.

The 18.09 section mentions NOT using the module now, due to the concerns about setcap, and instead suggests using it directly without the module. Both examples were shortened and made clearer.

If you could take another look, and offer your thoughts, I'd really appreciate it!
